### PR TITLE
UI: Add hotkey to save screenshot of preview

### PR DESCRIPTION
**Pull request by zavitax**
Adds the ability to save screenshots of the preview to PNG via hotkey.

### CODE REVIEW

File: UI/window-basic-settings.hpp [hunk 1]
```diff
@@ -170,6 +171,7 @@ class OBSBasicSettings : public QDialog {
 		outputsChanged = false;
 		audioChanged   = false;
 		videoChanged   = false;
+		screenshotsChanged	= false;
```
Comment (palana) [inline UI/window-basic-settings.hpp]: Found another case of weird formatting, you probably want a space instead of a tab here

File: UI/window-basic-main.cpp [hunk 2]
```diff
@@ -1603,6 +1619,301 @@ void OBSBasic::UpdateMultiviewProjectorMenu()
...
+		codec_context->pix_fmt, 32) < 0)
+	{
+		goto err_av_image_alloc;
+	}
+
+	// Copy image data, converting RGBA to
+	// image format expected by PNG encoder
+	swsc =
+		sws_getContext(
+			frame->width,
```
Comment (palana) [inline UI/window-basic-main.cpp]: So the overall formatting, while satisfying the line size limit just feels off a bit (sorry, and not sure if @jp9000 would mention this in his review), i.e. it's density is really low; I'd suggest something like
```
	swsc = sws_getContext(frame->width, frame->height, AV_PIX_FMT_RGBA,
			frame->width, frame->height, codec_context->pix_fmt,
			0, NULL, NULL, NULL);
```
semi-perhaps-logical grouping of parameters, so function calls don't span too many lines

Comment (zavitax) [reply UI/window-basic-main.cpp]: I disagree. Breaking args into one-per-line leaves space for comments.

**Comment by palana** [comment]
re:
> FYI you can rebase/squash commits and force-push to revise a PR. That's what is preferred for this project.

Yep, ideally some time* after the initial reviews are done

File: UI/forms/OBSBasicSettings.ui [hunk 3]
```diff
@@ -38,7 +38,7 @@
        </property>
        <property name="maximumSize">
         <size>
-         <width>135</width>
+         <width>165</width>
```
Comment (palana) [inline UI/forms/OBSBasicSettings.ui]: Another small thing, was it necessary to change the width?

Comment (zavitax) [reply UI/forms/OBSBasicSettings.ui]: Yes, otherwise "Screenshots" would not fit in the list horizontally

Comment (palana) [reply UI/forms/OBSBasicSettings.ui]: Alright

File: UI/window-basic-main.cpp (hunk 2 above)
Comment (palana) [reply UI/window-basic-main.cpp]: Sure, but there are no comments, so it's wasted space at this point

**Comment by OsirisNL** [comment]
Those first 2 commits should not be in this PR.

File: UI/window-basic-main.cpp [hunk 4]
```diff
@@ -1603,6 +1621,314 @@ void OBSBasic::UpdateMultiviewProjectorMenu()
...
+
+	// Save acquired image data to PNG
+	bool success = SaveRGBABufferToPNGFile(
+		outputFilePath,
+		callback_data.width,
+		callback_data.height,
+		callback_data.image_data_ptr,
+		callback_data.image_data_buffer_length,
+		callback_data.image_data_linesize
+	);
```
Comment (jp9000) [inline UI/window-basic-main.cpp]: Inconsistent code styling with the rest of the project when calling this function.  Should be:

```C
	bool success = SaveRGBABufferToPNGFile(
			outputFilePath,
			callback_data.width,
			callback_data.height,
			callback_data.image_data_ptr,
			callback_data.image_data_buffer_length,
			callback_data.image_data_linesize);
```

File: UI/window-basic-main.cpp [hunk 5]
```diff
@@ -1807,13 +2133,31 @@ void OBSBasic::CreateHotkeys()
...
+		{
+			OBSBasic &basic = *static_cast<OBSBasic*>(data);
+
+			if (pressed)
+			{
+				basic.SaveScreenshotHandler();
+			}
+		},
+		this);
+
```
Comment (jp9000) [inline UI/window-basic-main.cpp]: Code styling inconsistencies again.

To fix, you could do something like this:

```cpp
	takeScreenShotHotkey = obs_hotkey_register_frontend(
			"OBSBasic.SaveMainOutputScreenShot",
			Str("Basic.Main.SaveMainOutputScreenShot"),
			[] (void *data, obs_hotkey_id, obs_hotkey_t*,
					bool pressed)
			{
				OBSBasic &basic = *static_cast<OBSBasic*>(data);
                
				if (pressed) {
					basic.SaveScreenshotHandler();
				}
			},
			this);
```

Or if you want to make it a bit less squishy and clean looking:

```cpp
	auto screenshot_cb = [] (void *data, obs_hotkey_id, obs_hotkey_t*,
			bool pressed)
	{
		OBSBasic &basic = *static_cast<OBSBasic*>(data);

		if (pressed) {
			basic.SaveScreenshotHandler();
		}
	};

	takeScreenShotHotkey = obs_hotkey_register_frontend(
			"OBSBasic.SaveMainOutputScreenShot",
			Str("Basic.Main.SaveMainOutputScreenShot"),
			screenshot_cb,
			this);
```

File: UI/window-basic-main.cpp [hunk 6]
```diff
@@ -1807,13 +2133,31 @@ void OBSBasic::CreateHotkeys()
 			"OBSBasic.Transition",
 			Str("Transition"), transition, this);
 	LoadHotkey(transitionHotkey, "OBSBasic.Transition");
+
+	takeScreenShotHotkey = obs_hotkey_register_frontend(
+		"OBSBasic.SaveMainOutputScreenShot",
+		Str("Basic.Main.SaveMainOutputScreenShot"),
+		[](void* data, obs_hotkey_id, obs_hotkey_t*, bool pressed)
```
Comment (jp9000) [inline UI/window-basic-main.cpp]: Should be `void *data`.

File: UI/window-basic-settings.cpp [hunk 7]
```diff
@@ -3357,6 +3378,19 @@ void OBSBasicSettings::on_simpleOutputBrowse_clicked()
 	ui->simpleOutputPath->setText(dir);
 }
 
+void OBSBasicSettings::on_screenshotOutputBrowse_clicked()
+{
+	QString dir = QFileDialog::getExistingDirectory(this,
+		QTStr("Basic.Settings.ScreenShots.SelectDirectory"),
+		ui->screenshotOutputPath->text(),
+		QFileDialog::ShowDirsOnly |
+		QFileDialog::DontResolveSymlinks);
```
Comment (jp9000) [inline UI/window-basic-settings.cpp]: Inconsistent code syling with the rest of the project:

```cpp
	QString dir = QFileDialog::getExistingDirectory(this,
			QTStr("Basic.Settings.ScreenShots.SelectDirectory"),
			ui->screenshotOutputPath->text(),
			QFileDialog::ShowDirsOnly |
			QFileDialog::DontResolveSymlinks);
```

File: UI/window-basic-main.cpp [hunk 8]
```diff
@@ -1603,6 +1621,314 @@ void OBSBasic::UpdateMultiviewProjectorMenu()
...
+		// Signal the calling thread we're done
+		os_event_signal(args->complete_event);
+	};
+
+	// Add raw video callback to get the next video frame
+	obs_add_raw_video_callback(
+		&raw_video_format, // requested video format
+		callback_data.raw_video_callback_ref, // callback
+		&callback_data // callback arguments
+	);
```
Comment (jp9000) [inline UI/window-basic-main.cpp]: Trailing whitespace on this line.  Also, Inconsistent code styling with the rest of the project when calling this function. Should be:

```C
	obs_add_raw_video_callback(
			&raw_video_format,
			callback_data.raw_video_callback_ref,
			&callback_data);
```

File: UI/window-basic-main.cpp [hunk 9]
```diff
@@ -1603,6 +1621,314 @@ void OBSBasic::UpdateMultiviewProjectorMenu()
...
+	// initialize completion event
+	os_event_init(&callback_data.complete_event, OS_EVENT_TYPE_AUTO);
+
+	// initialize callback function input
+	callback_data.width = video_info.base_width;
+	callback_data.height = video_info.base_height;
+
+	// set up our callback function body for obs_add_raw_video_callback()
+	callback_data.raw_video_callback_ref =
+		[](void *param, video_data* data)
```
Comment (jp9000) [inline UI/window-basic-main.cpp]: Should be `video_data *data`.

File: UI/window-basic-main.cpp [hunk 10]
```diff
@@ -1603,6 +1621,314 @@ void OBSBasic::UpdateMultiviewProjectorMenu()
...
+			args->image_data_buffer_length);
+
+		// Remove our callback as soon as we're done
+		obs_remove_raw_video_callback(
+			args->raw_video_callback_ref,
+			param);
+
+		// Signal the calling thread we're done
+		os_event_signal(args->complete_event);
+	};
```
Comment (jp9000) [inline UI/window-basic-main.cpp]: Inconsistent indentation with the rest of the project.  I'll give you an example of how it should look normally:

```C
	callback_data.raw_video_callback_ref = [] (void *param,
			video_data *data)
	{
		raw_video_callback_frame_data_args *args =
				(raw_video_callback_frame_data_args *)param;

		// Copy screenshot data aside
		args->image_data_linesize = data->linesize[0];
		args->image_data_buffer_length =
				args->image_data_linesize * args->height;

		// Allocate memory for video data
		args->image_data_ptr = (uint8_t*)bmalloc(
				args->image_data_buffer_length);

		// Copy video data to our own buffer
		memcpy(args->image_data_ptr, data->data[0],
				args->image_data_buffer_length);

		// Remove our callback as soon as we're done
		obs_remove_raw_video_callback(
				args->raw_video_callback_ref,
				param);

		// Signal the calling thread we're done
		os_event_signal(args->complete_event);
	};
```

File: UI/window-basic-main.cpp [hunk 11]
```diff
@@ -1807,13 +2133,31 @@ void OBSBasic::CreateHotkeys()
...
+
+	takeScreenShotHotkey = obs_hotkey_register_frontend(
+		"OBSBasic.SaveMainOutputScreenShot",
+		Str("Basic.Main.SaveMainOutputScreenShot"),
+		[](void* data, obs_hotkey_id, obs_hotkey_t*, bool pressed)
+		{
+			OBSBasic &basic = *static_cast<OBSBasic*>(data);
+
+			if (pressed)
+			{
```
Comment (jp9000) [inline UI/window-basic-main.cpp]: Should be `if (pressed) {`

File: UI/window-basic-main.cpp [hunk 12]
```diff
@@ -1603,6 +1621,314 @@ void OBSBasic::UpdateMultiviewProjectorMenu()
...
+// @note				used by OBSBasic::SaveScreenshotToFile
+//
+static bool SaveRGBABufferToPNGFile(
+	const char *outputFilePath,
+	uint32_t    width,
+	uint32_t    height,
+	uint8_t*    image_data_ptr,
+	size_t      image_data_buffer_length,
+	uint32_t    image_data_linesize)
+{
```
Comment (jp9000) [inline UI/window-basic-main.cpp]: This indentation of the parameters is inconsistent with the rest of the project.  It should be more like this:

```cpp
static bool SaveRGBABufferToPNGFile(
		const char  *outputFilePath,
		uint32_t    width,
		uint32_t    height,
		uint8_t     *image_data_ptr,
		size_t      image_data_buffer_length,
		uint32_t    image_data_linesize)
{
```

**Comment by jp9000** [comment]
Additionally, your commit messages don't follow the commit message guideline.  This is what your commit message looks like raw:

```
UI: Save screenshot of main output to PNG
Add screenshot settings section
Add screenshot output directory setting
Add save output screenshot hotkey binding
Add swscale swresample libs to UI CMakeLists.txt
Use existing ffmpeg bindings to save to PNG
Use obs_add_raw_video_callback to get video frame
```

This is how is should look more ideally:

```
UI: Save screenshot of main output to PNG

Adds the ability to save screenshots of the preview to PNG via hotkey.
```

First line should be no more than 50 characters (excluding prefix), second line should be blank, and the rest of the lines are used for the detailed yet concise description of the commit (word wrapping at 72 characters).

File: UI/platform-osx.mm [hunk 13]
```diff
@@ -84,6 +84,21 @@ string GetDefaultVideoSavePath()
...
 }
 
+string GetDefaultScreenShotsSavePath()
+{
+	NSFileManager *fm = [NSFileManager defaultManager];
+	NSURL *url = [fm URLForDirectory:NSPicturesDirectory
+				inDomain:NSUserDomainMask
+		       appropriateForURL:nil
+				  create:true
+				   error:nil];
```
[the remark this answers is not in the document]
Comment (zavitax) [reply UI/platform-osx.mm]: This is correct. Looking at string GetDefaultVideoSavePath() above, we have the same indentation issues (I just copied & pasted the implementation as-is to keep them as similar as possible). I'll fix indentation in both.

File: UI/window-basic-main.cpp [hunk 14]
```diff
@@ -1603,6 +1621,314 @@ void OBSBasic::UpdateMultiviewProjectorMenu()
...
+	frame->width = width;
+	frame->height = height;
+
+	if (av_image_alloc(
+		&frame->data[0],                  // dest buffer
+		&frame->linesize[0],              // dest linesize (stride)
+		codec_context->width,             // dest width
+		codec_context->height,            // dest height
+		codec_context->pix_fmt,           // dest pixel format
+		32) < 0)                          // alignment
```
[the remark this answers is not in the document]
Comment (zavitax) [reply UI/window-basic-main.cpp]: For the comments part, I do see how they aid readability

**Comment by dodgepong** [comment]
Something that @palana just pointed out to me is that the screenshots are of the downscaled ouptut, re-upscaled to the canvas size. For example, if I have a 1920x1080 canvas resolution with a downscale to 640x360, the screenshot function will create a 1920x1080 image of the upscaled 640x360 output:

**Comment by ethaniel** [comment]
Little user-side request here.

Once this change is approved, can it please be made accessible via the websockets API?

I'm running a remote camera on a beach and using obs to broadcast it to youtube. The camera (hero 6) sometimes turns itself off and I have no way of knowing this. Right now I'm using the "replay buffer" which I'm triggering via websockets API every 10 seconds. I then have to use ffmpeg to extract keyframes from the resulting file which takes it's toll on the limited resources of the tiny computer that I'm using. (i compare the keyframes to each other, and if they're the same, then I reboot the camera).

It would be much easier and less CPU intensive to trigger a snapshot via websockets API directly.

**Comment by zavitax** [comment]
@dodgepong can you provide steps to reproduce?

**Comment by jp9000** [comment]
@zavitax Ah, it's because it uses obs_add_raw_video_callback().  obs_add_raw_video_callback() gets the output post-scaling.  (E.G., in video settings, set the "Canvas/Base Resolution" to 1920x1080, then set "Output/Scaled Resolution" to 1280x720 -- the screenshot code will only be able to get an image for the latter size, not the former)

I now realize in retrospect that may have been a mistake -- there may need to be a custom render solution for getting a screenshot after all.

**Comment by zavitax** [comment]
@jp9000 is it possible to add a pre-scale version of obs_add_raw_video_callback() ?

**Comment by jp9000** [comment]
It wouldn't make sense to add that.  It would only be used for the screenshot code alone.

**Comment by ntoff** [comment]
>I still found it disconcerting that there is no indication anywhere that a screenshot has been saved. I understand why we don't want an audio indicator, as it would be recorded when users might not want it recorded. Maybe a visual indicator, such as a temporary "Screenshot Saved" message in the status bar? Or a balloon notification? (are balloon notifications still a thing?) Obviously this is the sort of thing that could also be included in a game capture overlay, but that is an entirely different discussion.

I mean, there's no indication OBS saves *anything*, really. Save the replay buffer? Nothing at all to indicate it's been saved. No audio, no visual, no overlay, nothing. I hacked some functionality into the websockets plugin + obs tablet remote to allow me some replay buffer control and at least made the save button (in tablet remote) change to "saving ..." while saving, but that's still just a best guess that it did infact save based I think on perhaps the fact OBS didn't spit out an error (I'd have to go back and check what I actually did).

>It was a bit weird not seeing a setting for the Screenshot hotkey under the Screenshot settings tab.

I made a bit of a post on the forum about the hotkeys being all sorts of jumbled up or separated in weird ways. Like the replay buffer hot key being in a section all on its own for *saving* yet the buttons for starting and stopping it are up the top in with the streaming controls. OBS's hotkey section needs a major tidy up / rework. I tried to have a go but couldn't figure out how to rearrange them properly, I did find where they're hiding but never managed to properly rearrange them.

**Comment by zavitax** [comment]
@dodgepong @jp9000 @palana the scaling issue should be fixed now.

Could you test from your end, please?

File: UI/platform-osx.mm [hunk 15]
```diff
@@ -73,10 +73,25 @@ string GetDefaultVideoSavePath()
 {
 	NSFileManager *fm = [NSFileManager defaultManager];
 	NSURL *url = [fm URLForDirectory:NSMoviesDirectory
-				inDomain:NSUserDomainMask
-		       appropriateForURL:nil
-				  create:true
-				   error:nil];
+			inDomain:NSUserDomainMask
```
Comment (palana) [inline UI/platform-osx.mm]: Looks like you included changes which don't belong in this commit (and PR)

File: UI/window-basic-main.cpp [hunk 16]
```diff
@@ -1782,7 +2158,7 @@ void OBSBasic::CreateHotkeys()
 	LoadHotkeyPair(streamingHotkeys,
 			"OBSBasic.StartStreaming", "OBSBasic.StopStreaming");
 
-	auto cb = [] (void *data, obs_hotkey_id, obs_hotkey_t*, bool pressed)
+	auto cb = [](void *data, obs_hotkey_id, obs_hotkey_t*, bool pressed)
```
Comment (palana) [inline UI/window-basic-main.cpp]: There's some unrelated whitespace change here that doesn't belong in this commit

File: UI/window-basic-main.cpp [hunk 17]
```diff
@@ -1648,6 +1666,364 @@ void OBSBasic::UpdateMultiviewProjectorMenu()
...
+		}
+	};
+
+	// Add tick callback
+	obs_add_tick_callback(callback_data.tick_callback_ref,
+			&callback_data);
+
+	// Wait for our callback to be fired and for video frame
+	// to become available
+	os_event_wait(callback_data.complete_event);
```
Comment (palana) [inline UI/window-basic-main.cpp]: You're blocking any callers of `SaveScreenshotToFile` for multiple frames, i.e. the UI thread will wait for multiple graphics thread frames; this really shouldn't happen imo, kind of the same problem as the initial design that blocked the graphics thread instead

File: UI/platform-osx.mm (hunk 15 above)
Comment (zavitax) [reply UI/platform-osx.mm]: this is styling (whitespace) change requested by Jim in the previous set of comments

Comment (palana) [reply UI/platform-osx.mm]: It's not, he requested styling to change on the lines you added in that commit, not existing lines

Comment (zavitax) [reply UI/platform-osx.mm]: when you see a broken window, you fix it?

Comment (palana) [reply UI/platform-osx.mm]: By breaking git history? Nope

**Comment by ntoff** [comment]
>The only primary issue remaining is mostly just where to place the screenshot browse button.

In the "File" menu? Near the recording remux functions? I mean, that's where the "show settings folder" clickable menu item is so it makes sense all the other "browse x folder" items would also be there, no? yes?

**Comment by ethaniel** [comment]
> I want to add a suggestion; What about adding the ability to automatically take screenshots every X seconds?

I second that. I have a script which runs diagnostics (uploads thumbnails to youtube, calculates image "niceness", uses AI to count people). Right now I have to use the replay buffer and extract snapshots with ffmpeg. It would make my life much easier if OBS could automatically make a snapshots every X seconds.

**Comment by ethaniel** [comment]
Any updates?

**Comment by jd4no** [comment]
Any updates to this being merged? This PR would save me a ton of time.

**Comment by Bluscream** [comment]
Can i get a compiled version of this PR please?

**Comment by ethaniel** [comment]
@jp9000 when can this be allowed in?

**Comment by synap5e** [comment]
@ethaniel My plugin might be what you are after
https://github.com/synap5e/obs-screenshot-plugin
It lets you save/PUT screenshots from a single source at a given interval.

**Comment by yuya495** [comment]
I want the ability to take screenshots like xsplit

**Comment by GXTX** [comment]
@zavitax Do you have any will to continue work on this PR? It's unfortunate that it's been bikeshed to death here.

**Comment by zavitax** [comment]
> @zavitax Do you have any will to continue work on this PR? It's unfortunate that it's been bikeshed to death here.

Perhaps as a separate plug-in. What did you have in mind?

**Comment by jp9000** [comment]
Superseded by https://github.com/obsproject/obs-studio/pull/3202
